### PR TITLE
Sites Management: Abstract site list processing in a single component

### PR DESCRIPTION
**Pull request by zaguiini**
#### Proposed Changes

See https://github.com/Automattic/wp-calypso/issues/68043.

Preamble:

1. Moves all sites-related pieces of logic to a new packages called `@automattic/sites` and renames the "sites table" suffix to "sites list", which is more appropriate;
2. Splits `useSitesTableFiltering` into `useSitesListFiltering` and `useSitesListGrouping`.

This PR adds a new factory function called `createSitesListComponent`, which returns a new component that processes the `sites` prop, an array of sites, with pre-defined processing functions. This is done to maintain consistency between usages of the sites list across the application (i.e. site switcher, sites page). [The tests describe the factory and the intention following its usage](https://github.com/Automattic/wp-calypso/pull/68175/files#diff-8f180599ff89988931e6de53ff67db0cfc952be406ce28801d29c690e0414d7a).

The developer can disable pipes on demand by passing a `features` argument, an object. Example:

```ts
const SitesList = createSitesListComponent( {
    grouping: false
} )
```

Like that, the usage of `SitesList` won't split the sites by groups:

```tsx
<SitesList sites={originalSites} filtering={ { search: 'myterm' } }>
    {( { sites } ) => sites.map( renderSite )}
</SitesList>
```

The new factory function is used on the sites page (all pipes enabled) and the site switcher (except grouping). It is still possible to use the pipes independently by using its original hooks, such as useSitesListFiltering.

#### Testing Instructions

Check that there are no behavior changes on the sites page, nor the sites switcher usages across the application.

### CODE REVIEW

File: client/state/sites/hooks/with-sites-sorting.tsx [hunk 1]
```diff
@@ -2,10 +2,12 @@ import { createHigherOrderComponent } from '@wordpress/compose';
 import { ComponentType } from 'react';
 import { useSitesSorting } from './use-sites-sorting';
 
+export type SitesSortingPreferenceProps = ReturnType< typeof useSitesSorting >;
```
Comment (zaguiini) [inline client/state/sites/hooks/with-sites-sorting.tsx:5]: `useSitesSorting` will be renamed to `useSitesSortingPreference` in a follow-up PR to maintain consistency.

File: packages/sites/src/create-sites-list-component.ts [hunk 2]
```diff
@@ -0,0 +1,120 @@
+/* eslint-disable @typescript-eslint/ban-types, react-hooks/rules-of-hooks */
```
Comment (zaguiini) [inline packages/sites/src/create-sites-list-component.ts:1]: Looping @Automattic/team-calypso-frameworks to get a second opinion on how to achieve this.

Goals:

1. Allow specifying which hooks should be executed;
2. Reflect types as such (i.e. if `grouping` is disabled, do not expose a `statuses` property in the render child and do not allow passing a `grouping` property specifying grouping options in the usage).

Example:

```tsx
const SitesList = createSitesListComponent({ grouping: false })

<SitesList
  grouping={{ ... }} // this should error
>
  {({ sites, statuses /* this should error */ }) => sites}
</SitesList>
```

If `grouping` is `true`, then:

```tsx
const SitesList = createSitesListComponent({ grouping: false })

<SitesList
  grouping={{ ... }} // this should pass
>
  {({ sites, statuses /* this should pass */ }) => sites}
</SitesList>
```

**Comment by wojtekn** [comment]
Nice refactor!

> Check that there are no behavior changes on the sites page, nor the sites switcher usages across the application.

It works as expected for me. I needed to merge the trunk to this feature branch on my local to have some already merged features back.

**Review by wojtekn — APPROVED** [review]

**Review by danielbachhuber — APPROVED** [review]
Works fine in my testing!

Given the breadth of this refactor, I'm going to land the PR so it doesn't block the other things we're working on, and leave the issue open to address the API question.

**Pull request merged**